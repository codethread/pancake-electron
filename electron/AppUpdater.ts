import log from 'electron-log';
import { autoUpdater } from 'electron-updater';

export class AppUpdater {
  constructor() {
    log.transports.file.level = 'debug';
    autoUpdater.logger = log;
    autoUpdater.checkForUpdatesAndNotify().catch((err) => log.error(err));
  }
}
