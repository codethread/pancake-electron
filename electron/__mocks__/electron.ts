export const app = {
  getName: jest.fn(),
  getVersion: jest.fn(),
  quit: jest.fn(),
};

export const dialog = {
  showMessageBox: jest.fn().mockResolvedValue(null),
};
