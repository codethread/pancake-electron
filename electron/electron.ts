export {
  BrowserWindow,
  IpcRenderer,
  IpcMainInvokeEvent,
  IpcMainEvent,
  IpcMain,
  app,
  contextBridge,
  ipcMain,
  ipcRenderer,
  shell,
} from 'electron'; // eslint-disable-line import/no-extraneous-dependencies
