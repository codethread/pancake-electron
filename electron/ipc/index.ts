export { setupIpcHandlers } from './setupIpcHandlers';
export { handlers } from './handlers';
