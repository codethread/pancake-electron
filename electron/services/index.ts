export * from './logger';
export * from './updater';
export * from './devTools';
