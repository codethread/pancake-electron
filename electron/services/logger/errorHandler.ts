import { app, dialog } from 'electron';
import { CatchErrorsOptions } from 'electron-log';
import type { ILogger } from './createLogger';

export const errorHandler: (
  logger: ILogger
) => CatchErrorsOptions['onError'] = (logger) => (
  error,
  versions,
  submitIssue
) => {
  dialog
    .showMessageBox({
      title: 'An error occurred',
      message: error.message,
      detail: error.stack,
      type: 'error',
      buttons: ['Ignore', 'Report', 'Exit'],
    })
    .then((result) => {
      if (result.response === 1) {
        const body = generateReport(error, versions);

        submitIssue?.(
          'https://github.com/AHDesigns/pancake-electron/issues/new',
          {
            title: 'Error report',
            body,
            labels: 'to refine, bug',
          }
        );
        return;
      }
      if (result.response === 2) {
        app.quit();
      }
    })
    .catch(logger.errorWithContext('error submitting issue'));
};

function generateReport(
  error: Error,
  versions?: {
    os: string;
    electron: string;
    app: string;
  }
): string {
  const appVersion = versions?.app ?? 'no app version';
  const electron = versions?.electron ?? 'no electron version';
  const os = versions?.os ?? 'no OS detected';
  const code = '```';
  const heading = '**Error Report:**';
  const stack = error.stack
    ? `Stacktrace: \n${code}\n${error.stack}\n${code}`
    : 'No stacktrace';

  return `**Describe the bug**
Describe as best you can when the error occured, with steps leading up to it if possible.
**Screenshots**
If possible, add screenshots to help explain your problem.
${heading}
${stack}
OS: ${os}
${electron}
${appVersion}`;
}
