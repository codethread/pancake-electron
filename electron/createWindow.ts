import { BrowserWindow } from 'electron';
import path from 'path';
import url from 'url';
import { isDev, isIntegration } from '@shared/constants';
import { ILogger } from './services';

export function createWindow(logger: ILogger): BrowserWindow {
  /**
   * These options should be off in production and dev mode
   * They are only required for e2e testing with Spectron
   *
   * enableRemoteModule needs to be true to allow spectron to access information
   * about the running electron process
   *
   * contextIsolation needs to be disabled to give the browserWindow access to
   * require, which in turn is used to access `remote` as needed above
   */
  const secureEnvs = isIntegration
    ? {
        enableRemoteModule: true,
        contextIsolation: false,
      }
    : {
        enableRemoteModule: false,
        contextIsolation: true,
      };

  logger.debug(secureEnvs);

  const mainWindow = new BrowserWindow({
    width: 1100,
    height: 700,
    backgroundColor: '#191622',
    webPreferences: {
      ...secureEnvs,
      nodeIntegration: false,
      preload: path.join(__dirname, './preload.js'),
    },
  });

  if (isDev) {
    mainWindow.webContents.openDevTools();

    mainWindow
      .loadURL('http://localhost:4000')
      .catch(logger.errorWithContext('loading window from dev server'));
  } else {
    mainWindow
      .loadURL(
        url.format({
          pathname: path.join(__dirname, 'renderer/index.html'),
          protocol: 'file:',
          slashes: true,
        })
      )
      .catch(logger.errorWithContext('loading window from file'));
  }

  mainWindow.webContents.on('did-finish-load', () => {
    mainWindow.setTitle('Pancake');
  });

  return mainWindow;
}
