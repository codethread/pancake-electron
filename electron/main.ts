// eslint-disable-next-line import/no-extraneous-dependencies
import url from 'url';
import path from 'path';
import { checkForUpdates, createLogger, setUpDevtools } from '@electron/services';
import { ipcMain, app, BrowserWindow } from '@electron/electron';
import { productionRepositories } from '@electron/repositories';
import { fakeRepositories } from '@electron/repositories/fakes';
import { handlers, setupIpcHandlers } from '@electron/ipc';
import { isDev, isIntegration } from '@shared/constants';

import log from 'electron-log';
import { createWindow } from './windows/createWindow';

let mainWindow: BrowserWindow | null;
const logger = createLogger(log);

checkForUpdates(logger);
const repos = isIntegration ? fakeRepositories() : productionRepositories({ logger });
app
  .on('ready', () => {
    logger.info('app ready');
    mainWindow = createWindow(logger);

    mainWindow.on('closed', closeWindow);
  })
  .whenReady()
  .then(() => setupIpcHandlers(ipcMain, handlers(repos)))
  .then(() => setUpDevtools(logger))
  .catch(logger.errorWithContext('main window creation'));

function closeWindow(): void {
  mainWindow = null;
  logger.info('User closed window');
  app.quit();
}

// globalShortcut.register('Ctrl+Alt+P', () => {
//   // more https://stackoverflow.com/questions/50642126/previous-window-focus-electron if windows and linux don't play ball
//   if (mb.window?.isVisible()) {
//     mb.hideWindow();
//     mb.app.hide();
//   } else {
//     mb.showWindow();
//   }
// });
// });

// mb.on('will-quit', () => {
// globalShortcut.unregisterAll();
// });
