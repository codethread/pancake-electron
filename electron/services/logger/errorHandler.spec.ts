import * as electron from 'electron';
import { mocked } from 'ts-jest/utils';
import { logger } from './index';
import { errorHandler } from './errorHandler';

jest.mock('./index');
const { app, dialog } = mocked(electron, true);

describe('errorHandler', () => {
  const error = new Error('poop');
  error.stack = 'error stack for poop';
  const defaultVersions = { os: 'mac', electron: '1', app: 'pancake' };
  let versions: typeof defaultVersions | undefined;
  let submitIssue: jest.Mock | undefined = jest.fn();

  beforeEach(() => {
    const appliedErrorHandler = errorHandler(logger);
    appliedErrorHandler?.(error, versions, submitIssue);
  });

  it('creates a message box for the user', () => {
    expect(dialog.showMessageBox).toHaveBeenCalledWith({
      title: 'An error occurred',
      message: error.message,
      detail: error.stack,
      type: 'error',
      buttons: ['Ignore', 'Report', 'Exit'],
    });
  });

  describe('when the user chooses to ignore the error', () => {
    beforeAll(() => {
      dialog.showMessageBox.mockResolvedValue({
        response: 3,
        checkboxChecked: false,
      });
    });

    it('does nothing', () => {
      expect(submitIssue).not.toHaveBeenCalled();
      expect(app.quit).not.toHaveBeenCalled();
      expect(logger.errorWithContext('')).not.toHaveBeenCalled();
    });
  });

  describe('when the user chooses to quit', () => {
    beforeAll(() => {
      dialog.showMessageBox.mockResolvedValue({
        response: 2,
        checkboxChecked: false,
      });
    });

    it('does not log a report', () => {
      expect(submitIssue).not.toHaveBeenCalled();
      expect(logger.errorWithContext('')).not.toHaveBeenCalled();
    });

    it('quits the app', () => {
      expect(app.quit).toHaveBeenCalled();
    });
  });

  describe('when the user chooses to report the error', () => {
    beforeAll(() => {
      dialog.showMessageBox.mockResolvedValue({
        response: 1,
        checkboxChecked: false,
      });
    });

    it('does not quit the app', () => {
      expect(logger.errorWithContext('')).not.toHaveBeenCalled();
      expect(app.quit).not.toHaveBeenCalled();
    });

    describe('and all details are provided by electron', () => {
      beforeAll(() => {
        versions = defaultVersions;
      });

      it('opens a github issue', () => {
        expect(submitIssue).toHaveBeenCalledWith(
          'https://github.com/AHDesigns/pancake-electron/issues/new',
          /* eslint-disable @typescript-eslint/no-unsafe-assignment */
          {
            title: expect.any(String),
            body: expect.any(String),
            labels: 'to refine, bug',
          }
          /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        );
        expect(submitIssue?.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('and no details are provided by electron', () => {
      beforeAll(() => {
        versions = undefined;
      });

      it('opens a github issue', () => {
        expect(submitIssue).toHaveBeenCalledWith(
          'https://github.com/AHDesigns/pancake-electron/issues/new',
          /* eslint-disable @typescript-eslint/no-unsafe-assignment */
          {
            title: expect.any(String),
            body: expect.any(String),
            labels: 'to refine, bug',
          }
          /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        );
        expect(submitIssue?.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('and no stack trace is provided', () => {
      beforeAll(() => {
        versions = undefined;
        error.stack = undefined;
      });

      it('opens a github issue', () => {
        expect(submitIssue).toHaveBeenCalledWith(
          'https://github.com/AHDesigns/pancake-electron/issues/new',
          /* eslint-disable @typescript-eslint/no-unsafe-assignment */
          {
            title: expect.any(String),
            body: expect.stringContaining('No stacktrace'),
            labels: 'to refine, bug',
          }
          /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        );
        expect(submitIssue?.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('when no submit function is provided', () => {
      beforeAll(() => {
        submitIssue = undefined;
      });

      it('does nothing', () => {
        expect(app.quit).not.toHaveBeenCalled();
        expect(logger.errorWithContext('')).not.toHaveBeenCalled();
      });

      afterAll(() => {
        submitIssue = jest.fn();
      });
    });

    describe('when submit issue fails', () => {
      const err = new Error('poop');
      beforeAll(() => {
        dialog.showMessageBox.mockRejectedValue(err);
      });

      it('does not quit', () => {
        expect(app.quit).not.toHaveBeenCalled();
      });

      it('logs the exception', () => {
        expect(logger.errorWithContext('')).toHaveBeenCalledWith(err);
      });
    });
  });
});
