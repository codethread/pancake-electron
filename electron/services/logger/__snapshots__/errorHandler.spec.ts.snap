// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorHandler when the user chooses to report the error and all details are provided by electron opens a github issue 1`] = `
Array [
  "https://github.com/AHDesigns/pancake-electron/issues/new",
  Object {
    "body": "**Describe the bug**
Describe as best you can when the error occured, with steps leading up to it if possible.
**Screenshots**
If possible, add screenshots to help explain your problem.
**Error Report:**
Stacktrace: 
\`\`\`
error stack for poop
\`\`\`
OS: mac
1
pancake",
    "labels": "to refine, bug",
    "title": "Error report",
  },
]
`;

exports[`errorHandler when the user chooses to report the error and no details are provided by electron opens a github issue 1`] = `
Array [
  "https://github.com/AHDesigns/pancake-electron/issues/new",
  Object {
    "body": "**Describe the bug**
Describe as best you can when the error occured, with steps leading up to it if possible.
**Screenshots**
If possible, add screenshots to help explain your problem.
**Error Report:**
Stacktrace: 
\`\`\`
error stack for poop
\`\`\`
OS: no OS detected
no electron version
no app version",
    "labels": "to refine, bug",
    "title": "Error report",
  },
]
`;

exports[`errorHandler when the user chooses to report the error and no stack trace is provided opens a github issue 1`] = `
Array [
  "https://github.com/AHDesigns/pancake-electron/issues/new",
  Object {
    "body": "**Describe the bug**
Describe as best you can when the error occured, with steps leading up to it if possible.
**Screenshots**
If possible, add screenshots to help explain your problem.
**Error Report:**
No stacktrace
OS: no OS detected
no electron version
no app version",
    "labels": "to refine, bug",
    "title": "Error report",
  },
]
`;
